hal/src/nRF52840/spi_hal.cpp

```diff
@@ -448,20 +449,36 @@ void hal_spi_transfer_dma(hal_spi_interface_t spi, void* tx_buffer, void* rx_buf
         return;
     }
 
+    uint8_t* txPtr = nullptr;
+
+    SCOPE_GUARD ({
+        if (!nrfx_is_in_ram(tx_buffer)) {
+            free(txPtr);
+        }
+    });
+
+    if (!nrfx_is_in_ram(tx_buffer)) {
+        txPtr = (uint8_t*)malloc(length);
```

Comment:
I think we can do better. The whole idea with sending constant data out of flash is not to waste RAM. This approach pretty much negates it. What if there is a rather larger blob resource that needs to be transmitted e.g. > 50K?

We should be able to get away with a rather small stack/heap buffer (e.g. 256 bytes) and just send the data in multiple chunks. This shouldn't cause any problems, as SPI data is anyway clocked by the master and as long as the CS is held.

With SPI Slave mode, it doesn't seem that this would work, unless there was support for double-buffering. So for slave transfers we can probably keep the current approach or not allow out of RAM buffers.

Reply:
Will refactor it a little bit.

Reply:
I just realized that if we truncated the tx buffer, the following chunks except the first chunk cannot be sent, since the `spiTransfer` is not a blocking API and it just sets the tx/rx buffer and starts transmitting and then returns.

Reply:
We need to introduce some static variables to handle event properly for truncated data transmission.

Reply:
Oh, yeah, good catch. We need to add some state management for cases where we are performing a chunked transfer.